// soxjs2/lib/device_meta.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _xml2js = require('xml2js');

var _xml2js2 = _interopRequireDefault(_xml2js);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// import XmlUtil from './xml_util';

var _xmlDeclarePatStr = "^<\\?xml[^>]+?>";
var _xmlDeclarePat = new RegExp(_xmlDeclarePatStr);

var DeviceMeta = function () {
  function DeviceMeta(device, deviceId, deviceType, serialNumber, metaTransducers) {
    _classCallCheck(this, DeviceMeta);

    this.device = device;
    this.deviceId = deviceId;
    this.deviceType = deviceType;
    this.serialNumber = serialNumber;
    this.metaTransducers = metaTransducers;
  }

  _createClass(DeviceMeta, [{
    key: 'getXmlAttrs',
    value: function getXmlAttrs() {
      return {
        id: this.deviceId,
        type: this.deviceType,
        serialNumber: this.serialNumber,
        name: this.device.getName(),
        xmlns: 'http://jabber.org/protocol/sox'
      };
    }
  }, {
    key: 'getDevice',
    value: function getDevice() {
      return this.device;
    }
  }, {
    key: 'getName',
    value: function getName() {
      return this.getDevice().getName();
    }
  }, {
    key: 'getId',
    value: function getId() {
      return this.deviceId;
    }
  }, {
    key: 'getType',
    value: function getType() {
      return this.deviceType;
    }
  }, {
    key: 'getSerialNumber',
    value: function getSerialNumber() {
      return this.serialNumber;
    }
  }, {
    key: 'getMetaTransducers',
    value: function getMetaTransducers() {
      return this.metaTransducers;
    }
  }, {
    key: '_getContentForXmlBuild',
    value: function _getContentForXmlBuild() {
      // build content for xml2js.Builder
      var tMetas = this.metaTransducers.map(function (mtv) {
        return mtv._getContentForXmlBuild();
      });
      return {
        device: {
          '$': {
            xmlns: 'http://jabber.org/protocol/sox',
            name: this.getName(),
            id: this.deviceId,
            type: this.deviceType,
            serialNumber: this.serialNumber
          },
          transducer: tMetas
        }
      };
    }
  }, {
    key: 'toXmlString',
    value: function toXmlString() {
      // import XmlUtil from './xml_util';
      var builder = new _xml2js2.default.Builder({ renderOpts: { pretty: false } });
      var content = this._getContentForXmlBuild();
      var rawXmlStr = builder.buildObject(content);

      // remove <?xml ....?>
      // let trimmedXmlStr = XmlUtil.removeXmlDeclaration(rawXmlStr);
      var trimmedXmlStr = rawXmlStr.replace(_xmlDeclarePat, "");

      return trimmedXmlStr;
    }
  }, {
    key: 'appendToNode',
    value: function appendToNode(node) {
      // used when publish
      var ret = node.c('device', this.getXmlAttrs());

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.metaTransducers[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var tdr = _step.value;

          ret.c('transducer', tdr.getXmlAttrs()).up();
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return ret;
    }
  }]);

  return DeviceMeta;
}();

module.exports = DeviceMeta;
