// soxjs2/lib/device.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Device = function () {
  function Device(soxConnection, deviceName, domain) {
    _classCallCheck(this, Device);

    if (domain === undefined) {
      domain = soxConnection.getDomain();
    }

    this.soxConnection = soxConnection;
    this.name = deviceName;
    this.domain = domain;
  }

  _createClass(Device, [{
    key: "getName",
    value: function getName() {
      return this.name;
    }
  }, {
    key: "getDomain",
    value: function getDomain() {
      return this.domain;
    }
  }, {
    key: "getMetaNodeName",
    value: function getMetaNodeName() {
      return this.name + "_meta";
    }
  }, {
    key: "getDataNodeName",
    value: function getDataNodeName() {
      return this.name + "_data";
    }
  }, {
    key: "getBoundSoxConnection",
    value: function getBoundSoxConnection() {
      return this.soxConnection;
    }
  }, {
    key: "addListener",
    value: function addListener(callback, listenerId) {
      var conn = this.getBoundSoxConnection();
      conn.addListeer(this, callback, listenerId);
    }
  }, {
    key: "removeAllListener",
    value: function removeAllListener() {
      var conn = this.getBoundSoxConnection();
      conn.removeAllListenerForDevice(this);
    }
  }, {
    key: "fetchMeta",
    value: function fetchMeta(callback) {
      var conn = this.getBoundSoxConnection();
      conn.fetchMeta(this, callback);
    }
  }, {
    key: "publish",
    value: function publish(data) {
      var conn = this.getBoundSoxConnection();
      return conn.publish(self, data);
    }
  }]);

  return Device;
}();

module.exports = Device;
