// soxjs2/lib/meta_transducer.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MetaTransducer = function () {
  function MetaTransducer(device, name, tdrId, _canActuate, _hasOwnNode, units, unitScalar, minValue, maxValue, resolution) {
    _classCallCheck(this, MetaTransducer);

    this.device = device;
    this.name = name;
    this.tdrId = tdrId;
    this._canActuate = _canActuate;
    this._hasOwnNode = _hasOwnNode;
    this.units = units;
    this.unitScalar = unitScalar;
    this.minValue = minValue;
    this.maxValue = maxValue;
    this.resolution = resolution;
  }

  _createClass(MetaTransducer, [{
    key: 'getDevice',
    value: function getDevice() {
      return this.device;
    }
  }, {
    key: 'getName',
    value: function getName() {
      return this.name;
    }
  }, {
    key: 'getId',
    value: function getId() {
      return this.tdrId;
    }
  }, {
    key: 'canActuate',
    value: function canActuate() {
      return this._canActuate;
    }
  }, {
    key: 'hasOwnNode',
    value: function hasOwnNode() {
      return this._hasOwnNode;
    }
  }, {
    key: 'getUnits',
    value: function getUnits() {
      return this.units;
    }
  }, {
    key: 'getUnitScalar',
    value: function getUnitScalar() {
      return this.unitScalar;
    }
  }, {
    key: 'getMinValue',
    value: function getMinValue() {
      return this.minValue;
    }
  }, {
    key: 'getMaxValue',
    value: function getMaxValue() {
      return this.maxValue;
    }
  }, {
    key: 'getResolution',
    value: function getResolution() {
      return this.resolution;
    }
  }, {
    key: 'getXmlAttrs',
    value: function getXmlAttrs() {
      var attrs = {
        name: this.name,
        id: this.tdrId
        // canActuate: String(this.canActuate()),
        // hasOwnNode: String(this.hasOwnNode()),
        // units: this.units,
        // unitScalar: this.unitScalar,
        // minValue: this.minValue,
        // maxValue: this.maxValue,
        // resolution: this.resolution
      };

      if (this._canActuate !== undefined) {
        attrs.canActuate = String(this.canActuate());
      }

      if (this._hasOwnNode !== undefined) {
        attrs.hasOwnNode = String(this.hasOwnNode());
      }

      if (this.units) {
        attrs.units = this.units;
      }

      if (this.unitScalar) {
        attrs.unitScalar = this.unitScalar;
      }

      if (this.minValue) {
        attrs.minValue = this.minValue;
      }

      if (this.maxValue) {
        attrs.maxValue = this.maxValue;
      }

      if (this.resolution) {
        attrs.resolution = this.resolution;
      }

      return attrs;
    }
  }, {
    key: '_getContentForXmlBuild',
    value: function _getContentForXmlBuild() {
      // build content for xml2js.Builder()
      return { '$': this.getXmlAttrs() };
    }
  }]);

  return MetaTransducer;
}();

module.exports = MetaTransducer;
