// soxjs2/lib/sox_connection.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _nodeStrophe = require("node-strophe");

var _nodeStrophe2 = _interopRequireDefault(_nodeStrophe);

var _xml2js = require("xml2js");

var _xml2js2 = _interopRequireDefault(_xml2js);

var _sox_util = require("./sox_util");

var _sox_util2 = _interopRequireDefault(_sox_util);

var _xml_util = require("./xml_util");

var _xml_util2 = _interopRequireDefault(_xml_util);

var _device = require("./device");

var _device2 = _interopRequireDefault(_device);

var _transducer_value = require("./transducer_value");

var _transducer_value2 = _interopRequireDefault(_transducer_value);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Strophe = _nodeStrophe2.default.Strophe;

var $pres = Strophe.$pres;
var $iq = Strophe.$iq;

var PUBSUB_NS = "http://jabber.org/protocol/pubsub";
var PUBSUB_OWNER_NS = "http://jabber.org/protocol/pubsub#owner";

var SoxConnection = function () {
  function SoxConnection(boshService, jid, password) {
    _classCallCheck(this, SoxConnection);

    this.boshService = boshService;
    this.jid = jid;
    this.password = password;

    this._rawConn = null;
    this._isConnected = false;
    this._dataCallbacks = {};
    this._metaCallbacks = {};

    this._connEventCallbacks = {};
  }

  _createClass(SoxConnection, [{
    key: "_stropheOnRawInput",
    value: function _stropheOnRawInput(data) {
      //console.log("<<<<<< input");
      //console.log(data);
    }
  }, {
    key: "_stropheOnRawOutput",
    value: function _stropheOnRawOutput(data) {
      //console.log(">>>>>> output");
      //console.log(data);
    }
  }, {
    key: "addConnectionEventListner",
    value: function addConnectionEventListner(listener, listenerId) {
      if (listenerId === undefined) {
        listenerId = this._genRandomId();
      }

      this._connEventCallbacks[listenerId] = listener;
      return listenerId;
    }
  }, {
    key: "_callConnEvent",
    value: function _callConnEvent(methodName) {
      var callbacks = this._connEventCallbacks;
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = Object.keys(callbacks)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var callbackId = _step.value;

          var listener = callbacks[callbackId];
          var callback = listener[methodName];
          try {
            if (callback === undefined) {
              console.warn('callbackId=' + callbackId + " has not such method: " + methodName);
            } else {
              callback();
            }
          } catch (e) {
            console.error(e);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "_stropheOnConnConnecting",
    value: function _stropheOnConnConnecting() {
      this._callConnEvent('onConnecting');
    }
  }, {
    key: "_stropheOnConnConnected",
    value: function _stropheOnConnConnected() {
      // console.log("connected 1");
      this._rawConn.send($pres().c('priority').t('-1'));
      // console.log("### connected 2");

      // this._rawConn.PubSub.bind(
      //   "xmpp:pubsub:last-published-item",
      //   that._onLastPublishedItemReceived
      // );

      // this._rawConn.PubSub.bind(
      //   "xmpp:pubsub:item-published",
      //   that._onPublishedItemReceived
      // );

      var that = this;

      var pubsubHandler = function pubsubHandler(ev) {
        // TODO
        try {
          // console.log('@@@@@ pubsubHandler!');
          // XmlUtil.dumpDom(ev);
          var cb = function cb(data) {
            // console.log("@@@@@ got data!");
          };
          var data = _sox_util2.default.parseDataPayload(that, ev, cb);
          // TODO: dispatch
          that.dispatchData(data);
        } catch (ex) {
          console.error(ex);
        }
        return true; // needed to be called every time
      };

      var service = 'pubsub.' + this.getDomain();

      this._rawConn.addHandler(pubsubHandler, null, 'message', null, null, service);

      this._isConnected = true;
      // console.log("### connected 3");
      if (this._onConnectCallback) {
        // console.log("### connected 3-1");
        this._onConnectCallback();
        // console.log("### connected 3-2");
      }
      // console.log("### connected 4 end");
      this._callConnEvent('onConnected');
    }
  }, {
    key: "_stropheOnConnDisconnecting",
    value: function _stropheOnConnDisconnecting() {
      this._callConnEvent('onDisconnecting');
    }
  }, {
    key: "_stropheOnConnDisconnected",
    value: function _stropheOnConnDisconnected() {
      this._rawConn = null;
      this._isConnected = false;
      if (this._onDisconnectCallback) {
        this._onDisconnectCallback();
      }
      this._callConnEvent('onDisconnected');
    }
  }, {
    key: "_stropheOnConnFaill",
    value: function _stropheOnConnFaill() {
      this._callConnEvent('onFail');
    }
  }, {
    key: "_stropheOnConnectionStatusUpdate",
    value: function _stropheOnConnectionStatusUpdate(status) {
      // console.log("@@ start of _stropheOnConnectionStatusUpdate");
      if (status === Strophe.Strophe.Status.CONNECTING) {
        // console.log("@@connecting");
        this._stropheOnConnConnecting();
      } else if (status === Strophe.Strophe.Status.CONNFAIL) {
        // console.log("@@connfail");
        this._stropheOnConnFaill();
      } else if (status === Strophe.Strophe.Status.DISCONNECTING) {
        // console.log("@@disconnecting");
        this._stropheOnConnDisconnecting();
      } else if (status === Strophe.Strophe.Status.DISCONNECTED) {
        // console.log("@@disconnected");
        this._stropheOnConnDisconnected();
      } else if (status === Strophe.Strophe.Status.CONNECTED) {
        // console.log("@@connected");
        this._stropheOnConnConnected();
      } else {}
      // console.log("@@ UNKNOWN STATUS: " + status);

      // console.log("@@ end of _stropheOnConnectionStatusUpdate");
      return true;
    }

    // _stropheOnLastPublishedItemReceived(obj) {
    //   let node = obj.node;
    //   if (SoxUtil.endsWithMeta(node)) {
    //     this.dispatchMetaPublish(obj);
    //   } else if (SoxUtil.endsWithData(node)) {
    //     this.dispatchDataPublish(obj);
    //   } else {
    //     // FIXME
    //   }
    // }

    // _stropheOnPublishedItemReceived(obj) {
    //   let node = obj.node;
    //   if (SoxUtil.endsWithData(node)) {
    //     this.dispatchDataPublish(obj);
    //   } else {
    //     // FIXME
    //   }
    // }

    // dispatchDataPublish(obj) {
    //   let node = obj.node;
    //   let deviceName = SoxUtil.cutDataSuffix(node);
    //   let deviceListenerTable = this._dataCallbacks[deviceName];
    //   if (deviceListenerTable === undefined) {
    //     return;
    //   }
    //
    //   let deviceToBind = this.bind(deviceName);
    //   let that = this;
    //   let onDataParsed = (data) => {
    //     that._broadcast(deviceListenerTable, data);
    //   };
    //   SoxUtil.parseDataPayload(obj.entry, deviceToBind, onDataParsed);
    //   // this._broadcast(deviceListenerTable, data);
    // }

  }, {
    key: "dispatchData",
    value: function dispatchData(data) {
      var deviceName = data.getDevice().getName();
      var dataListenerTable = this._dataCallbacks[deviceName];
      if (dataListenerTable === undefined) {
        return;
      }

      this._broadcast(dataListenerTable, data);
    }

    // dispatchMetaPublish(obj) {
    //   let node = obj.node;
    //   let deviceName = SoxUtil.cutMetaSuffix(node);
    //   let deviceListenerTable = this._metaCallbacks[deviceName];
    //   if (deviceListenerTable === undefined) {
    //     return;
    //   }
    //
    //   let deviceToBind = this.bind(deviceName);
    //   let that = this;
    //   let onMetaParsed = (meta) => {
    //     that._broadcast(deviceListenerTable, meta);
    //   };
    //   SoxUtil.parseMetaPayload(obj.entry, deviceToBind, onMetaParsed);
    //   // let meta = SoxUtil.parseMetaPayload(obj.entry, deviceToBind);
    //   // this._broadcast(deviceListenerTable, meta);
    // }

  }, {
    key: "getBoshService",
    value: function getBoshService() {
      return this.boshService;
    }
  }, {
    key: "getDomain",
    value: function getDomain() {
      return Strophe.Strophe.getDomainFromJid(this.getJID());
    }
  }, {
    key: "getJID",
    value: function getJID() {
      return this.jid;
    }
  }, {
    key: "getPassword",
    value: function getPassword() {
      return this.password;
    }
  }, {
    key: "connect",
    value: function connect(callback) {
      var conn = new Strophe.Strophe.Connection(this.getBoshService());
      this._onConnectCallback = callback;
      conn.rawInput = this._stropheOnRawInput;
      conn.rawOutput = this._stropheOnRawOutput;
      this._rawConn = conn;
      var jid = this.getJID();
      var password = this.getPassword();

      // without wrapping call of _stropheOnConnectionStatusUpdate, "this" will be missed inside the func
      var that = this;
      var cb = function cb(status) {
        return that._stropheOnConnectionStatusUpdate(status);
      };
      conn.connect(jid, password, cb);
    }
  }, {
    key: "disconnect",
    value: function disconnect(callback) {
      if (this._rawConn !== null && this.isConnected()) {
        this._onDisconnectCallback = callback;
        this._rawConn.disconnect();
      }
    }
  }, {
    key: "isConnected",
    value: function isConnected() {
      return this._isConnected;
    }
  }, {
    key: "getStropheConnection",
    value: function getStropheConnection() {
      return this._rawConn;
    }
  }, {
    key: "addListener",
    value: function addListener(device, callback, listenerId) {
      if (listenerId === undefined) {
        listenerId = this._genRandomId();
      }
      this._registerDataListener(device, listenerId, callback);
      return listenerId;
    }
  }, {
    key: "removeAllListenerForDevice",
    value: function removeAllListenerForDevice(device) {
      this._dataCallbacks = {};
    }
  }, {
    key: "removeListener",
    value: function removeListener(listenerId) {
      this._removeDataListenerWithId(listenerId);
    }
  }, {
    key: "fetchMeta",
    value: function fetchMeta(device, callback) {
      var _this = this;

      try {
        var that = this;
        var listenerId = this._genRandomId();
        var metaNode = device.getMetaNodeName();
        var _callback = function _callback(meta) {
          that._unsubNode(device.getMetaNodeName(), device.getDomain(), function () {});
          callback(meta);
        };
        var service = "pubsub." + this.getDomain();
        this._registerMetaListener(device, listenerId, _callback);

        var cb = function cb(subscriptions) {
          var jid = that._rawConn.jid;
          var mySub = subscriptions[jid];
          if (mySub !== undefined) {
            var metaNodeSubIDs = mySub[metaNode];
            var availableSubID = metaNodeSubIDs[0];

            var uniqueId = that._rawConn.getUniqueId("pubsub");
            var iq2 = $iq({ type: "get", from: jid, to: service, id: uniqueId }).c("pubsub", { xmlns: PUBSUB_NS }).c("items", { node: metaNode, max_items: 1, subid: availableSubID });
            var suc2 = function suc2(iq) {
              // console.log("\n\nrecent request success?\n\n");
            };
            var err2 = function err2(iq) {
              // console.log("\n\nrecent request failed?\n\n");
            };
            that._rawConn.sendIQ(iq2, suc2, err2);do {} while (true);
          } else {
            // first we need to sub
            // console.log("\n\n\n@@@@@ no our sub info, going to sub!\n\n\n");
            var rawJid = _this._rawConn.jid;
            var bareJid = Strophe.Strophe.getBareJidFromJid(_this._rawConn.jid);
            var subIq = $iq({ to: service, type: "set", id: _this._rawConn.getUniqueId("pubsub") }).c('pubsub', { xmlns: "http://jabber.org/protocol/pubsub" }).c('subscribe', { node: metaNode, jid: rawJid });

            var subSuc = function subSuc(iq) {
              // console.log("\n\n@@@@ sub success, going to fetch subscriptions to get subid");
              that._getSubscription(device.getMetaNodeName(), device.getDomain(), function (subscriptions2) {
                var mySub2 = subscriptions2[jid];
                var metaNodeSubIDs2 = mySub2[metaNode];
                var availableSubID2 = metaNodeSubIDs2[0];

                var uniqueId3 = that._rawConn.getUniqueId("pubsub");
                var iq3 = $iq({ type: "get", from: jid, to: service, id: uniqueId3 }).c("pubsub", { xmlns: PUBSUB_NS }).c("items", { node: metaNode, max_items: 1, subid: availableSubID2 });

                var suc3 = function suc3(iq) {
                  var meta = _xml_util2.default.convRecentItem(that, iq);
                  _callback(meta);
                };
                var err3 = function err3(iq) {
                  // console.log("\n\n@@@@@ recent request error? 3\n\n");
                };

                that._rawConn.sendIQ(iq3, suc3, err3);
              });
            };
            that._rawConn.sendIQ(subIq, subSuc, function () {});
          }
        };
        this._getSubscription(device.getMetaNodeName(), device.getDomain(), cb);
      } catch (e) {
        console.log(e.stack);
      }
    }
  }, {
    key: "_getSubscription",
    value: function _getSubscription(node, domain, cb) {
      // <iq type='get'
      //     from='francisco@denmark.lit/barracks'
      //     to='pubsub.shakespeare.lit'
      //     id='subscriptions1'>
      //   <pubsub xmlns='http://jabber.org/protocol/pubsub'>
      //     <subscriptions/>
      //   </pubsub>
      // </iq>
      var service = "pubsub." + domain;
      var uniqueId = this._rawConn.getUniqueId("pubsub");
      var iq = $iq({ type: "get", from: this._rawConn.jid, to: service, id: uniqueId }).c("pubsub", { xmlns: PUBSUB_NS }).c("subscriptions");

      var suc = function suc(iq) {
        var converted = _xml_util2.default.convSubscriptions(iq);
        cb(converted);
      };
      var err = function err(iq) {};

      this._rawConn.sendIQ(iq, suc, err);
    }
  }, {
    key: "bind",
    value: function bind(deviceName, domain) {
      if (domain === undefined) {
        domain = this.getDomain();
      }

      return new _device2.default(this, deviceName, domain);
    }
  }, {
    key: "fetchDevices",
    value: function fetchDevices(callback, domain) {
      if (domain === undefined) {
        domain = this.getDomain();
      }
      // https://github.com/strophe/strophejs-plugin-pubsub/blob/master/strophe.pubsub.js#L297
      var jid = this.getJID();
      var service = "pubsub." + domain;
      var iq = $iq({ from: jid, to: service, type: "get", id: this._rawConn.getUniqueId("pubsub") }).c('query', { xmlns: Strophe.Strophe.NS.DISCO_ITEMS });

      var that = this;
      var success = function success(msg) {
        var query = msg._childNodesList[0];
        var items = query._childNodesList;

        var check = {};
        for (var i = 0; i < items.length; i++) {
          var item = items[i];
          var node = item._attributes.node._valueForAttrModified;
          if (_sox_util2.default.endsWithData(node)) {
            var realNode = _sox_util2.default.cutDataSuffix(node);
            if (check[realNode] === undefined) {
              check[realNode] = { data: true };
            } else {
              check[realNode].data = true;
            }
          } else if (_sox_util2.default.endsWithMeta(node)) {
            var _realNode = _sox_util2.default.cutMetaSuffix(node);
            if (check[_realNode] === undefined) {
              check[_realNode] = { meta: true };
            } else {
              check[_realNode].data = true;
            }
          }
        }

        var devices = [];
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = Object.keys(check)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var deviceName = _step2.value;

            var c = check[deviceName];
            if (c.data && c.meta) {
              var device = that.bind(deviceName);
              devices.push(device);
            }
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }

        callback(devices);
      };

      var error = function error(msg) {};

      return this._rawConn.sendIQ(iq.tree(), success, error, undefined);
    }
  }, {
    key: "fetchSubscriptions",
    value: function fetchSubscriptions(callback) {
      this._rawConn.PubSub.getSubscriptions(function (subscriptions) {
        // TODO: Device オブジェクトのリストに加工してcallbackを呼び出す

      });
    }
  }, {
    key: "subscribe",
    value: function subscribe(device) {
      var dataNode = device.getDataNodeName();
      var domain = device.getDomain();
      // let service = "pubsub." + device.getDomain();

      // this._subNode(dataNode, device.getDomain());
      var that = this;

      this.unsubscribe(device, function () {
        // console.log("@@@ unsubscribe callback called");
        var cb = function cb() {};
        that._subNode(dataNode, domain, false, cb);
        // console.log("@@@ _subNode called");
      });
    }
  }, {
    key: "_subNode",
    value: function _subNode(node, domain, requestRecent, callback) {
      // https://github.com/strophe/strophejs-plugin-pubsub/blob/master/strophe.pubsub.js#L297
      var that = this;
      var service = "pubsub." + domain;

      // http://ggozad.com/strophe.plugins/docs/strophe.pubsub.html
      // console.log("@@@@@@@ raw jid = " + this._rawConn.jid);
      var rawJid = this._rawConn.jid;
      var bareJid = Strophe.Strophe.getBareJidFromJid(this._rawConn.jid);
      var iq = $iq({ to: service, type: "set", id: this._rawConn.getUniqueId("pubsub") }).c('pubsub', { xmlns: "http://jabber.org/protocol/pubsub" }).c('subscribe', { node: node, jid: rawJid });

      var suc = function suc(iq) {
        // https://xmpp.org/extensions/xep-0060.html#subscriber-retrieve-requestrecent

        // <iq type='get'
        //     from='francisco@denmark.lit/barracks'
        //     to='pubsub.shakespeare.lit'
        //     id='items2'>
        //   <pubsub xmlns='http://jabber.org/protocol/pubsub'>
        //     <items node='princely_musings' max_items='2'/>
        //   </pubsub>
        // </iq>
        if (requestRecent) {
          var uniqueId = that._rawConn.getUniqueId("pubsub");
          var iq2 = $iq({ type: "get", from: that._rawConn.jid, to: service, id: uniqueId }).c("pubsub", { xmlns: PUBSUB_NS }).c("items", { node: node, max_items: 1 });
          var suc2 = function suc2(iq) {
            if (callback) {
              callback();
            }
          };
          var err2 = function err2(iq) {};
          that._rawConn.sendIQ(iq2, suc2, err2);
        } else {
          callback();
        }
      };
      var err = function err(iq) {};
      this._rawConn.sendIQ(iq, suc, err);
    }
  }, {
    key: "unsubscribe",
    value: function unsubscribe(device, callback) {
      var dataNode = device.getDataNodeName();
      var domain = device.getDomain();
      var that = this;

      var cb = function cb() {
        if (callback) {
          callback();
        }
      };

      var myJid = Strophe.Strophe.getBareJidFromJid(this._rawConn.jid);

      this._getSubscription(dataNode, domain, function (sub) {
        // console.log("_getSubscription callback called in unsubscribe");
        if (sub[myJid] === undefined) {
          sub[myJid] = {};
        }
        var subids = sub[myJid][dataNode];
        if (subids === undefined) {
          // console.log("@@@ subids === undefined!");
          cb();
          return;
        }
        // console.log("@@@ subids.length===" + subids.length);
        if (subids.length == 0) {
          that._unsubNode(dataNode, domain, cb);
        } else {
          var delNextFunc = function delNextFunc(i) {
            if (subids.length <= i) {
              return cb;
            }
            return function () {
              that._unsubNode(dataNode, domain, delNextFunc(i + 1), subids[i]);
              // console.log("@@@ _unsubNode called for subid=" + subids[i]);
            };
          };

          that._unsubNode(dataNode, domain, delNextFunc(1), subids[0]);
          // console.log("@@@ _unsubNode called for subid=" + subids[0]);
        }
      });
      // this._unsubNode(dataNode, domain, () => {
      //   // TODO
      // });
    }
  }, {
    key: "_unsubNode",
    value: function _unsubNode(node, domain, callback, subid) {
      var service = "pubsub." + domain;
      // <iq type='set'
      // from='francisco@denmark.lit/barracks'
      // to='pubsub.shakespeare.lit'
      // id='unsub1'>
      //   <pubsub xmlns='http://jabber.org/protocol/pubsub'>
      //      <unsubscribe
      //          node='princely_musings'
      //          jid='francisco@denmark.lit'/>
      //   </pubsub>
      // </iq>
      var bareJid = Strophe.Strophe.getBareJidFromJid(this._rawConn.jid);
      // console.log("_unsubNode: bareJid=" + bareJid);

      var unsubAttrs = { node: node, jid: bareJid };
      if (subid !== undefined) {
        unsubAttrs.subid = subid;
      }

      var iq = $iq({ to: service, type: "set", id: this._rawConn.getUniqueId("pubsub") }).c('pubsub', { xmlns: "http://jabber.org/protocol/pubsub" }).c('unsubscribe', unsubAttrs);

      var suc = function suc(iq) {
        // console.log("unsub success");
        if (callback) {
          callback(iq);
        }
      };
      var err = function err(iq) {
        // console.log("unsub failed");
        // XmlUtil.dumpDom(iq);
      };
      this._rawConn.sendIQ(iq, suc, err);
    }
  }, {
    key: "unsubscribeAll",
    value: function unsubscribeAll() {
      var that = this;
      this.fetchSubscriptions(function (devices) {
        var _iteratorNormalCompletion3 = true;
        var _didIteratorError3 = false;
        var _iteratorError3 = undefined;

        try {
          for (var _iterator3 = devices[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            var device = _step3.value;

            that.unsubscribe(device);
          }
        } catch (err) {
          _didIteratorError3 = true;
          _iteratorError3 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion3 && _iterator3.return) {
              _iterator3.return();
            }
          } finally {
            if (_didIteratorError3) {
              throw _iteratorError3;
            }
          }
        }
      });
    }
  }, {
    key: "createDevice",
    value: function createDevice(device, meta, cbSuccess, cbFailed) {
      try {
        var domain = device.getDomain();
        var metaNode = device.getMetaNodeName();
        var dataNode = device.getDataNodeName();
        var that = this;
        this._createNode(metaNode, domain, function (iq) {
          that._createNode(dataNode, domain, function (iq2) {
            // TODO: send meta to meta node
            that._publishToNode(metaNode, device.getDomain(), meta, cbSuccess, cbFailed);
          }, cbFailed);
        }, cbFailed);
      } catch (e) {
        console.log(e.stack);
      }
    }
  }, {
    key: "_createNode",
    value: function _createNode(nodeName, domain, cbSuccess, cbFailed) {
      // console.log("\n\n---- _createNode");
      var service = 'pubsub.' + domain;
      var conn = this._rawConn;
      var uniqueId = conn.getUniqueId('pubsub');
      // console.log("\n\n---- _createNode2");
      try {
        // const iq = $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid })
        //   .c('pubsub', { xmlns: PUBSUB_NS })
        //   .c('create', { node: nodeName });
        var iq = $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid }).c('pubsub', { xmlns: PUBSUB_NS }).c('create', { node: nodeName }).c('configure').c('x', { xmlns: 'jabber:x:data', type: 'submit' }).c('field', { var: 'pubsub#access_model', type: 'list-single' }).c('value').t('open').up().up().c('field', { var: 'pubsub#publish_model', type: 'list-single' }).c('value').t('open').up().up().c('field', { var: 'pubsub#persist_items', type: 'boolean' }).c('value').t('1').up().up().c('field', { var: 'pubsub#max_items', type: 'text-single' }).c('value').t('1');
        // console.log("\n\n---- _createNode3");

        conn.sendIQ(iq, cbSuccess, cbFailed);
        // console.log("\n\n---- _createNode4");
      } catch (e) {
        console.log(e.stack);
      }
    }
  }, {
    key: "_deleteNode",
    value: function _deleteNode(nodeName, domain, cbSuccess, cbFailed) {
      var service = 'pubsub.' + domain;
      var conn = this._rawConn;
      var uniqueId = conn.getUniqueId('pubsub');
      // const bareJid = Strophe.Strophe.getBareJidFromJid(conn.jid);
      // const fromJid = conn.
      var iq =
      // const iq = $iq({ to: service, type: 'set', id: uniqueId, from: bareJid })
      $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid }).c('pubsub', { xmlns: PUBSUB_OWNER_NS }).c('delete', { node: nodeName });

      conn.sendIQ(iq, cbSuccess, cbFailed);
    }
  }, {
    key: "deleteDevice",
    value: function deleteDevice(device, cbSuccess, cbFailed) {
      var domain = device.getDomain();
      var metaNode = device.getMetaNodeName();
      var dataNode = device.getDataNodeName();
      var that = this;
      this._deleteNode(metaNode, domain, function (iq) {
        that._deleteNode(dataNode, domain, cbSuccess, cbFailed);
      }, function (iq) {
        cbFailed(iq);
        that._deleteNode(dataNode, domain, function (iq2) {}, function (iq2) {});
      });
    }
  }, {
    key: "publish",
    value: function publish(data, cbSuccess, cbFailed) {
      var device = data.getDevice();
      var domain = device.getDomain();
      var dataNode = device.getDataNodeName();
      this._publishToNode(dataNode, domain, data, cbSuccess, cbFailed);
    }
  }, {
    key: "_publishToNode",
    value: function _publishToNode(nodeName, domain, publishContent, cbSuccess, cbFailed) {
      // expects publishContent as an instance of DeviceMeta or Data
      try {
        var service = 'pubsub.' + domain;
        var conn = this._rawConn;
        var uniqueId = conn.getUniqueId('pubsub');
        var itemUniqueId = conn.getUniqueId('item');
        var iq = $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid }).c('pubsub', { xmlns: PUBSUB_NS }).c('publish', { node: nodeName }).c('item', { id: itemUniqueId })
        // .cnode(publishContent)
        ;

        publishContent.appendToNode(iq);

        conn.sendIQ(iq, cbSuccess, cbFailed);
      } catch (e) {
        console.error(e.stack);
      }
    }
  }, {
    key: "_genRandomId",
    value: function _genRandomId() {
      var chars = "abcdef01234567890";
      var nChars = chars.length;
      var len = 128;
      var ret = "";
      for (var i = 0; i < len; i++) {
        var idx = Math.floor(Math.random() * nChars);
        var char = chars.charAt(idx);
        ret = ret + char;
      }
      return ret;
    }
  }, {
    key: "_registerMetaListener",
    value: function _registerMetaListener(device, listenerId, callback) {
      this._registerListener(this._metaCallbacks, device, listenerId, callback);
    }
  }, {
    key: "_registerDataListener",
    value: function _registerDataListener(device, listenerId, callback) {
      this._registerListener(this._dataCallbacks, device, listenerId, callback);
    }
  }, {
    key: "_registerListener",
    value: function _registerListener(table, device, listenerId, callback) {
      var deviceName = device.getName();

      if (table[deviceName] === undefined) {
        table[deviceName] = {};
      }

      table[deviceName][listenerId] = callback;
    }
  }, {
    key: "_broadcast",
    value: function _broadcast(table, argument) {
      var _iteratorNormalCompletion4 = true;
      var _didIteratorError4 = false;
      var _iteratorError4 = undefined;

      try {
        for (var _iterator4 = Object.keys(table)[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
          var listenerId = _step4.value;

          var listener = table[listenerId];
          // console.log('$$$$ listenerId=' + listenerId + ", listener=" + listener);
          listener(argument);
        }
      } catch (err) {
        _didIteratorError4 = true;
        _iteratorError4 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion4 && _iterator4.return) {
            _iterator4.return();
          }
        } finally {
          if (_didIteratorError4) {
            throw _iteratorError4;
          }
        }
      }
    }
  }, {
    key: "_removeMetaListenerWithId",
    value: function _removeMetaListenerWithId(listenerId) {
      this._removeListenerWithId(this._metaCallbacks, listenerId);
    }
  }, {
    key: "_removeDataListenerWithId",
    value: function _removeDataListenerWithId(listenerId) {
      this._removeListenerWithId(this._dataCallbacks, listenerId);
    }
  }, {
    key: "_removeListenerWithId",
    value: function _removeListenerWithId(table, listenerId) {
      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        for (var _iterator5 = Object.keys(table)[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
          var devName = _step5.value;

          var devTable = table[devName];
          var found = false;
          var _iteratorNormalCompletion6 = true;
          var _didIteratorError6 = false;
          var _iteratorError6 = undefined;

          try {
            for (var _iterator6 = Object.keys(devTable)[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
              var lstnId = _step6.value;

              if (lstnId === listenerId) {
                found = true;
                break;
              }
            }
          } catch (err) {
            _didIteratorError6 = true;
            _iteratorError6 = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion6 && _iterator6.return) {
                _iterator6.return();
              }
            } finally {
              if (_didIteratorError6) {
                throw _iteratorError6;
              }
            }
          }

          if (found) {
            delete devTable[listenerId];
            break;
          }
        }
      } catch (err) {
        _didIteratorError5 = true;
        _iteratorError5 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion5 && _iterator5.return) {
            _iterator5.return();
          }
        } finally {
          if (_didIteratorError5) {
            throw _iteratorError5;
          }
        }
      }
    }
  }, {
    key: "setAccessPermission",
    value: function setAccessPermission(nodeName, domain, accessModel, cbSuccess, cbFailed) {
      try {
        var service = 'pubsub.' + domain;
        var conn = this._rawConn;
        var uniqueId = conn.getUniqueId('pubsub');

        var iq = $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid }).c('pubsub', { xmlns: PUBSUB_OWNER_NS }).c('configure', { node: nodeName }).c('x', { xmlns: 'jabber:x:data', type: 'submit' }).c('field', { var: 'pubsub#access_model', type: 'list-single' }).c('value').t(accessModel);
        conn.sendIQ(iq, cbSuccess, cbFailed);
      } catch (e) {
        console.error(e.stack);
      }
    }
  }, {
    key: "setAffaliation",
    value: function setAffaliation(nodeName, domain, affaliation, cbSuccess, cbFailed) {
      try {
        var service = 'pubsub.' + domain;
        var conn = this._rawConn;
        var uniqueId = conn.getUniqueId('pubsub');

        var iq = $iq({ to: service, type: 'set', id: uniqueId, from: conn.jid }).c('pubsub', { xmlns: PUBSUB_OWNER_NS }).c('affiliations', { node: nodeName });

        for (var i = 0; i < affaliation.length; i++) {
          iq.c('affiliation', { xmlns: PUBSUB_OWNER_NS, jid: affaliation[i], affiliation: 'none' }).up();
        }
        conn.sendIQ(iq, cbSuccess, cbFailed);
      } catch (e) {
        console.error(e.stack);
      }
    }
  }]);

  return SoxConnection;
}();

module.exports = SoxConnection;
