// soxjs2/lib/transducer_value.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var TransducerValue = function () {
  function TransducerValue(transducerId, rawValue, typedValue, timestamp) {
    _classCallCheck(this, TransducerValue);

    this.transducerId = transducerId;
    this.rawValue = rawValue;
    this.typedValue = typedValue;
    if (timestamp === undefined) {
      timestamp = new Date();
    }
    this.timestamp = timestamp;
  }

  _createClass(TransducerValue, [{
    key: 'getTransducerId',
    value: function getTransducerId() {
      return this.transducerId;
    }
  }, {
    key: 'getRawValue',
    value: function getRawValue() {
      return this.rawValue;
    }
  }, {
    key: 'getTypedValue',
    value: function getTypedValue() {
      return this.typedValue;
    }
  }, {
    key: 'getTimestamp',
    value: function getTimestamp() {
      return this.timestamp;
    }
  }, {
    key: '_getContentForXmlBuild',
    value: function _getContentForXmlBuild() {
      // build content for xml2js.Builder
      return { '$': this.getXmlAttrs() };
    }
  }, {
    key: 'getXmlAttrs',
    value: function getXmlAttrs() {
      var ts = this.timestamp.toISOString();
      return {
        'id': this.transducerId,
        'rawValue': this.rawValue,
        'typedValue': this.typedValue,
        'timestamp': ts
      };
    }
  }]);

  return TransducerValue;
}();

module.exports = TransducerValue;
